HW03g.sql

```diff
@@ -0,0 +1,176 @@
+﻿--1 Выберите сотрудников, которые являются продажниками, и еще не сделали ни одной продажи.
+SELECT
+*
+FROM Application.People p
+WHERE p.IsSalesperson = 1
+AND NOT EXISTS (SELECT * FROM Sales.Orders s WHERE s.SalespersonPersonID = p.PersonID )
+
+--2. Выберите товары с минимальной ценой (подзапросом), 2 варианта подзапроса. 
+SELECT
+s.StockItemName
+FROM Warehouse.StockItems s
+WHERE s.UnitPrice <= (SELECT MIN(UnitPrice) FROM Warehouse.StockItems )
+
+SELECT
+s.StockItemName
+FROM Warehouse.StockItems s
+WHERE s.UnitPrice <= ALL (SELECT UnitPrice FROM Warehouse.StockItems )
+
+--3. Выберите информацию по клиентам, которые перевели компании 5 максимальных платежей из [Sales].[CustomerTransactions] 
+--   представьте 3 способа (в том числе с CTE)
+SELECT
+sc.CustomerName
+,tab1.TransactionAmount
+,tab1.TransactionDate
+FROM Sales.Customers sc
+JOIN (
+SELECT TOP 5 WITH TIES
+	s.TransactionAmount
+	,s.CustomerID
+	,s.TransactionDate
+	FROM Sales.CustomerTransactions s
+	ORDER BY s.TransactionAmount DESC
+) as tab1 ON tab1.CustomerID = sc.CustomerID
+
+;WITH cteTopTr as
+(
+SELECT TOP 5 WITH TIES
+	s.TransactionAmount
+	,s.CustomerID
+	,s.TransactionDate
+	FROM Sales.CustomerTransactions s
+	ORDER BY s.TransactionAmount DESC
+)
+SELECT 
+c.CustomerName
+,cte.TransactionAmount
+,cte.TransactionDate
+FROM cteTopTr cte
+JOIN Sales.Customers c ON  cte.CustomerID = c.CustomerID;
+
+SELECT 
+(SELECT
+c.CustomerName
+FROM Sales.Customers c WHERE c.CustomerID = s.CustomerID) as CustomerName
+,s.TransactionAmount
+,s.TransactionDate
+FROM Sales.CustomerTransactions s
+WHERE s.TransactionAmount IN  
+	(SELECT TOP 5 WITH TIES
+	s.TransactionAmount
+	FROM Sales.CustomerTransactions s
+	ORDER BY s.TransactionAmount DESC)
+
+--4. Выберите города (ид и название), в которые были доставлены товары, 
+--   входящие в тройку самых дорогих товаров, а также Имя сотрудника, который осуществлял упаковку заказов
+
+;WITH
+cteTopExpStockItems AS
+(
+SELECT TOP 3 WITH TIES
+UnitPrice
+,StockItemID
+FROM Warehouse.StockItems s
+ORDER BY UnitPrice DESC
+),
+cteOL AS
+(
+SELECT DISTINCT
+OrderID
+FROM  Sales.OrderLines
+WHERE StockItemID IN (SELECT StockItemID FROM cteTopExpStockItems)
+)
+SELECT 
+c.DeliveryCityID
+,ac.CityName
+,ap.FullName
+,o.PickedByPersonID
+FROM Sales.Orders o
+LEFT JOIN Sales.Customers c ON c.CustomerID = o.CustomerID
```

Comment:
`CustomerID` в `Sales.Orders` всегда not null - можно использовать `JOIN`
```suggestion
JOIN Sales.Customers c ON c.CustomerID = o.CustomerID
```
Так сразу понятно, что вы не ожидаете там null